github.io documentation for aipc (#124)
opea-project/docs: examples/ChatQnA/deploy/aipc.md

# Single node on-prem deployment with Ollama on Xeon Scalable processors

> 

Reviewer: You didn't change the title (this isn't a Xeon Scalable processor doc it's an AI PC doc, right?

> 

Author: @dbkinder changed the title

> This deployment section covers single-node on-prem deployment of the ChatQnA

Author: @mkbhanda @dbkinder I was adhering to the Xeon template and didn't change the common sections such as this. I understand that this will apply to many other places in the APIC document. I can apply these changes forming a new template for AIPC if it's okay with you.

> This deployment section covers single-node on-prem deployment of the ChatQnA

Reviewer: Yes, we don't want to change the "common" sections at this point.  Right @preethivenkatesh @tomlenth

example with OPEA comps to deploy using Ollama. There are several

> slice-n-dice ways to enable RAG with vectordb and LLM models, but here we will

Author: @mkbhanda @dbkinder Same issue as above as I was adhering to a template.

::::{tab-set}

:::{tab-item} Ollama
:sync: Ollama

We set up the Ollama LLM service with one command
```bash
curl -fsSL https://ollama.com/install.sh | sh
```

Next, we'll build the Ollama microservice docker. This will set the entry point
needed for Ollama to suit the ChatQnA examples
```
docker build --no-cache -t opea/llm-ollama:latest --build-arg https_proxy=$https_proxy \
   --build-arg http_proxy=$http_proxy -f comps/llms/text-generation/ollama/langchain/Dockerfile .
```

Set Ollama Service Configuration

Ollama Service Configuration file is `/etc/systemd/system/ollama.service`.
Edit the file to set OLLAMA_HOST environment (Replace **${host_ip}** with your host IPV4).
```
Environment="OLLAMA_HOST=${host_ip}:11434"
```
Set https_proxy environment for Ollama if your system access network through proxy.
```
Environment="https_proxy=http://proxy.example.com:8080"
```
Restart Ollama services
```
$ sudo systemctl daemon-reload
$ sudo systemctl restart ollama.service
```

Pull LLM model

```bash
export OLLAMA_HOST=http://${host_ip}:11434
ollama pull llama3
ollama list

NAME            ID              SIZE    MODIFIED
llama3:latest   365c0bd3c000    4.7 GB  5 days ago
```

> 

Reviewer: Close off the tab-item 
```suggestion
:::
```

632a6634b06b   opea/llm-ollama                                         "bash entrypoint.sh"     29 seconds ago   Up 27 seconds               0.0.0.0:9000->9000/tcp, :::9000->9000/tcp                                              llm-ollama
```
:::
::::

## Interacting with ChatQnA deployment

This section will walk you through what are the different ways to interact with
the microservices deployed

### Dataprep Microservice（Optional）

If you want to add/update the default knowledge base, you can use the following
commands. The dataprep microservice extracts the texts from variety of data
sources, chunks the data, embeds each chunk using embedding microservice and
store the embedded vectors in the redis vector database.

Local File `nke-10k-2023.pdf` Upload:

```
curl -X POST "http://${host_ip}:6007/v1/dataprep" \
     -H "Content-Type: multipart/form-data" \
     -F "files=@./nke-10k-2023.pdf"
```

This command updates a knowledge base by uploading a local file for processing.
Update the file path according to your environment.

> Add Knowledge Base via HTTP Links:

Reviewer 2: ```suggestion
Upload from HTTP Links:
```

```
curl -X POST "http://${host_ip}:6007/v1/dataprep" \
     -H "Content-Type: multipart/form-data" \
     -F 'link_list=["https://opea.dev"]'
```

> This command updates a knowledge base by submitting a list of HTTP links for processing.

Reviewer 2: ```suggestion
This above command updates a knowledge store by submitting a list of HTTP/s links for processing.
```

> Also, you are able to get the file list that you uploaded:

Reviewer 2: ```suggestion
You can list all the data uploaded:
```

```
curl -X POST "http://${host_ip}:6007/v1/dataprep/get_file" \
     -H "Content-Type: application/json"

> 

Reviewer 2: Consider adding the output of a list command, in this case a file and na url link

```

> To delete the file/link you uploaded you can use the following commands:

Reviewer 2: ```suggestion
The command below enables you to delete the file/link you uploaded earlier:
```

     -d '{"file_path": "https://opea.dev.txt"}' \
     -H "Content-Type: application/json"
```

#### Delete file

```
curl -X POST "http://${host_ip}:6007/v1/dataprep/delete_file" \
     -d '{"file_path": "nke-10k-2023.pdf"}' \
     -H "Content-Type: application/json"
```

#### Delete all uploaded files and links

```
curl -X POST "http://${host_ip}:6007/v1/dataprep/delete_file" \
     -d '{"file_path": "all"}' \
     -H "Content-Type: application/json"
```
### TEI Embedding Service

The TEI embedding service takes in a string as input, embeds the string into a
vector of a specific length determined by the embedding model and returns this
embedded vector.

```
curl ${host_ip}:6006/embed \
    -X POST \
    -d '{"inputs":"What is Deep Learning?"}' \
    -H 'Content-Type: application/json'
```

In this example the embedding model used is "BAAI/bge-base-en-v1.5", which has a
vector size of 768. So the output of the curl command is a embedded vector of
length 768.

### Embedding Microservice
The embedding microservice depends on the TEI embedding service. In terms of
input parameters, it takes in a string, embeds it into a vector using the TEI

> embedding service and pads other default parameters that are required for the

Reviewer 2: ```suggestion
embedding service and adds other default parameters that are required for the
```

retrieval microservice and returns it.

```
curl http://${host_ip}:6000/v1/embeddings\
  -X POST \
  -d '{"text":"hello"}' \
  -H 'Content-Type: application/json'
```
### Retriever Microservice

To consume the retriever microservice, you need to generate a mock embedding

> vector by Python script. The length of embedding vector is determined by the

Reviewer 2: ```suggestion
vector using a Python script. The length of embedding vector is determined by the
```

```

### TEI Reranking Service

The TEI Reranking Service reranks the documents returned by the retrieval
service. It consumes the query and list of documents and returns the document
index based on decreasing order of the similarity score. The document
corresponding to the returned index with the highest score is the most relevant
document for the input query.
```
curl http://${host_ip}:8808/rerank \
    -X POST \
    -d '{"query":"What is Deep Learning?", "texts": ["Deep Learning is not...", "Deep learning is..."]}' \
    -H 'Content-Type: application/json'
```

Output is:  `[{"index":1,"score":0.9988041},{"index":0,"score":0.022948774}]`

### Reranking Microservice

The reranking microservice consumes the TEI Reranking service and pads the

> response with default parameters required for the llm microservice.

Reviewer 2: ```suggestion
response with default parameters required for the LLM microservice.
```

```
curl http://${host_ip}:8000/v1/reranking\
  -X POST \
  -d '{"initial_query":"What is Deep Learning?", "retrieved_docs": \
     [{"text":"Deep Learning is not..."}, {"text":"Deep learning is..."}]}' \
  -H 'Content-Type: application/json'
```

The input to the microservice is the `initial_query` and a list of retrieved
documents and it outputs the most relevant document to the initial query along

> with other default parameter such as temperature, `repetition_penalty`,

Reviewer 2: ```suggestion
with other default parameters such as the temperature, `repetition_penalty`,
```

{"model":"llama3","created_at":"2024-09-05T08:47:26.232468605Z","response":" learning","done":false}
{"model":"llama3","created_at":"2024-09-05T08:47:27.584534136Z","response":" that","done":false}
{"model":"llama3","created_at":"2024-09-05T08:47:28.50201424Z","response":" involves","done":false}
{"model":"llama3","created_at":"2024-09-05T08:47:29.895471763Z","response":" the","done":false}
{"model":"llama3","created_at":"2024-09-05T08:47:31.204128984Z","response":" use","done":false}
{"model":"llama3","created_at":"2024-09-05T08:47:32.231884525Z","response":" of","done":false}
{"model":"llama3","created_at":"2024-09-05T08:47:33.510913894Z","response":" artificial","done":false}
{"model":"llama3","created_at":"2024-09-05T08:47:34.516291108Z","response":" neural","done":false}
...
```

:::
::::

### LLM Microservice
```
curl http://${host_ip}:9000/v1/chat/completions\
  -X POST \
  -d '{"query":"What is Deep Learning?","max_new_tokens":17,"top_k":10,"top_p":0.95,\
     "typical_p":0.95,"temperature":0.01,"repetition_penalty":1.03,"streaming":true}' \
  -H 'Content-Type: application/json'

```

> You will get generated text from LLM:

Reviewer 2: ```suggestion
You will get the below generated text from the LLM:
```

```
chaqna-aipc-conversation-ui-server:
  image: opea/chatqna-conversation-ui:latest
  container_name: chatqna-aipc-conversation-ui-server
  environment:
    - APP_BACKEND_SERVICE_ENDPOINT=${BACKEND_SERVICE_ENDPOINT}
    - APP_DATA_PREP_SERVICE_URL=${DATAPREP_SERVICE_ENDPOINT}
  ports:
    - "5174:80"
  depends_on:
    - chaqna-aipc-backend-server
  ipc: host
  restart: always
```

Once the services are up, open the following URL in your browser: http://{host_ip}:5174. By default, the UI runs on port 80 internally. If you prefer to use a different host port to access the frontend, you can modify the port mapping in the compose.yaml file as shown below:

```
  chaqna-aipc-conversation-ui-server:
    image: opea/chatqna-conversation-ui:latest
    ...
    ports:
      - "80:80"
```

### Stop the services

Once you are done with the entire pipeline and wish to stop and remove all the containers, use the command below:
::::{tab-set}

:::{tab-item} Ollama
:sync: Ollama

```
docker compose -f compose.yaml down
```

> 

Reviewer: close off the tab-item
```suggestion
:::
```
